### PR TITLE
new worker-based api

### LINKED ISSUE #10: Alternative to string-based function
Filing this issue because it is in my todo list.

Reported:
- can't use external function #6 
- warning about inline function #8 

I'd like to find a way to make it more practical.
Maybe, it needs to cooperate with bundlers.

### PR DESCRIPTION
**Pull request by dai-shi**
This closes #10.

The new API doesn't allow string based functions.
It requires to create a worker outside of the hook.
This is to let bundlers to do the job.

It may not work with CRA. I will think about workaround.

### CODE REVIEW

**Comment by dai-shi** [comment]
@dangdennis @pocesar @ScriptedAlchemy
I would like to have some feedbacks from you guys.
Please see examples.
I have only tested with worker-plugin.

You can install this branch by `npm install dai-shi/react-hooks-worker#new-api-with-worker-plugin`

File: src/useWorker.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    worker.onmessageerror = () => dispatchSafe({ type: 'messageerror' });
+    const cleanup = () => {
+      dispatchSafe = () => null; // we should not dispatch after cleanup.
+      dispatch({ type: 'init' });
+    };
+    return cleanup;
+  }, [worker]);
+  useEffect(() => {
+    lastWorker.current.postMessage(input);
+  }, [input]);
```
Comment (pocesar) [inline src/useWorker.js]: missing `lastWorker` as deps

**Comment by pocesar** [comment]
it's nice, the API reminds of comlink. the only issue I see when initializing a Worker outside of components is that it'll be a global instance instead of a local instantiation. had the same problem when I was creating a react wrapper for comlink, so I made 3 APIs, for each use case (see https://github.com/pocesar/react-use-comlink). 
if you go with just one, i'd say, make it explicit in the readme. else you might want to pass in a flag to the hook to instantiate when it's first used inside the context of a different element (`new Worker` creates a instance, `() => new Worker` creates a factory)

**Comment by dai-shi** [comment]
Oh, it's very nice to see that you've already tried hooks with comlinkjs.
I thought about it too, but I'd rather want non-RPC style. (like examples/05_generator)

I wasn't confident if worker-plugin works for the factory pattern. I'd go with `() => new Worker()`.

File: src/useWorker.js (hunk 1 above)
Comment (dai-shi) [reply src/useWorker.js]: This is actually OK. lastWorker is made by useRef and is consistent. The exhaustive-deps rule doesn't complain it.

**Comment by dai-shi** [comment]
Changed to `createWorker` API.

> make it explicit in the readme

I'm not sure if I understand this point. Isn't it clean enough? (I'd admit my README is minimal, though.)

**Comment by pocesar** [comment]
bad phrasing on my part. I meant to make it clear that the same worker instance is used on all components before. if people is using external variables, the variables will keep state between components, that might be confusing. but now with a factory, that won't happen

**Comment by dai-shi** [comment]
Got it! Thanks for the explanation.

**Comment by ScriptedAlchemy** [comment]
Will review today! Thanks for thinking of me

**Comment by dai-shi** [comment]
As for CRA, I will wait for the official support. https://github.com/facebook/create-react-app/pull/5886

**Comment by dai-shi** [comment]
Published: https://www.npmjs.com/package/react-hooks-worker/v/0.8.0

**Pull request merged**